File: clex.scala
import java.io.InputStream

object CLex {

	var yyval:AST.AST = null
	var yytext:String = ""

	private var buf:Option[Char] = None
	private var in:InputStream = null

	def setInputStream(in:InputStream) {
		CLex.in = in
	}

	// for debug
	private def getChar():Char = {
		buf match {
		case None => in.read().asInstanceOf[Char]
		case Some(a) => buf = None; a
		}
	}

	private def ungetChar(c:Char) {
		buf = Some(c)
	}

	private def isspace(c:Char):Boolean = {
		c match {
		case ' ' | '\t' | '\r' | '\n' => true
		case _ => false
		}
	}

	private def isdigit(c:Char):Boolean = {
		if (c >= '0' && c <= '9') true else false
	}

	private def isalpha(c:Char):Boolean = {
		if (c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z') true else false
	}
	def yylex():Int = {
		var c:Char = ' '
		while (isspace(c)) {
			c = getChar()
		}

	    c match {
	    case '+' | '-' | '*' | '>' | '<' | '(' | ')' | '{' | '}' | ']' | '[' | ';' | ',' | '=' | EOF => c
	    case '"' => yytext = ""
	    	var flg = true
			while (flg) {
				val c = getChar()
				if (c == '"') {
					flg = false
				} else {
					yytext += c
				}
			}
			yyval = AST.STR(yytext)
			Parser.STRING
		case _ if (isdigit(c)) =>
			var n = 0
			do {
			    n = n * 10 + c - '0'
			    c = getChar()
			} while(isdigit(c))
			ungetChar(c)
			yyval = AST.NUM(n)
			Parser.NUMBER
		case _  if(isalpha(c)) =>
			yytext = ""
			do {
				yytext += c
				c = getChar()
			} while(isalpha(c))
			ungetChar(c)
			yytext match {
			case "var" => Parser.VAR
			case "if" => Parser.IF
			case "else" => Parser.ELSE
			case "return" => Parser.RETURN
			case "while" => Parser.WHILE
			case "for" => Parser.FOR
			case "println" => Parser.PRINTLN
			case _ => yyval = AST.makeSymbol(yytext); Parser.SYMBOL
			}
		case _ => throw new Exception("bad char '" + c + "'")
		}
	}
}

File: CLex.java
package stinyc;
import java.io.Reader;
import java.io.IOException;

class CLex {

	String yytext = "";
	static final int EOF = -1;
	private int buf = -2;
	private Reader in = null;
	private Parser parser;
	public CLex(Reader r, Parser p) {
		this.in = r;
		this.parser = p;
	}

	// for debug
	private int getChar() throws IOException {
		if (buf == -2) {
			int c = in.read();
			return c;
		} else {
			int buf1 = buf;
			buf = -2;
			return buf1;
		}
	}

	private void ungetChar(int c) {
		buf = c;
	}

	private boolean isspace(int c) {
		if (c == ' ' || c == '\t' || c == '\r' || c == '\n') return true;
		else return false;
	}

	private boolean isdigit(int c) {
		if (c >= '0' && c <= '9') return true;
		else return false;
	}

	private boolean isalpha(int c) {
		if (c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z') return true;
		else return false;
	}

	int yylex() throws IOException {
		int c = ' ';
		while (isspace(c)) {
			c = getChar();
		}

		switch (c) {
	    case '+':
		case '-':
		case '*':
		case '>':
		case '<':
		case '(':
		case ')':
		case '{':
		case '}':
		case ']':
		case '[':
		case ';':
		case ',':
		case '=':
		case EOF: return c;
		case '"': yytext = "";
			boolean flg = true;
			while (flg) {
				c = getChar();
				if (c == '"') {
					flg = false;
				} else {
					yytext += (char)c;
				}
			}
			parser.yylval = new ParserVal(yytext);
			return Parser.STRING;
		default:
			if (isdigit(c)) {
				int n = 0;
				do {
				    n = n * 10 + c - '0';
				    c = getChar();
				} while(isdigit(c));
				ungetChar(c);
				parser.yylval = new ParserVal(n);
				return Parser.NUMBER;
			} else if(isalpha(c)) {
				yytext = "";
				do {
					yytext += (char)c;
					c = getChar();
				} while(isalpha(c));
				ungetChar(c);
				if (yytext.equals("var")) return Parser.VAR;
				if (yytext.equals("if")) return Parser.IF;
				if (yytext.equals("else")) return Parser.ELSE;
				if (yytext.equals("return")) return Parser.RETURN;
				if (yytext.equals("while")) return Parser.WHILE;
				if (yytext.equals("for")) return Parser.FOR;
				if (yytext.equals("println")) return Parser.PRINTLN;
				parser.yylval = new ParserVal((Object)parser.makeSymbol(yytext));
				return Parser.SYMBOL;
				
			} else {
				throw new IOException("bad char '" + c + "'");
			}
		}
	}
}
